runtime/src/accounts_db.rs

```diff
@@ -1240,14 +1234,16 @@ impl AccountsDB {
         self.accounts_index.write().unwrap().add_root(slot)
     }
 
-    pub fn get_rooted_storage_entries(&self) -> Vec<Arc<AccountStorageEntry>> {
+    pub fn get_snapshot_storages(&self, snapshot_slot: Slot) -> SnapshotStorages {
         let accounts_index = self.accounts_index.read().unwrap();
         let r_storage = self.storage.read().unwrap();
         r_storage
             .0
-            .values()
-            .flat_map(|slot_store| slot_store.values().cloned())
-            .filter(|store| accounts_index.is_root(store.slot_id))
+            .iter()
+            .filter(|(slot, storage)| {
+                **slot <= snapshot_slot && accounts_index.is_root(**slot) && !storage.is_empty()
```

Comment:
`slot <= snapshot_slot` logic is moved from [here](https://github.com/solana-labs/solana/pull/8339/files#diff-0f0af8fcfa1444e1b07d671c5bfa4a69L89).